docs/proposals/improve-status-in-CAPI-resources.md

```diff
@@ -0,0 +1,958 @@
+
+```go
+type MachineSpec struct {
+
+    // MinReadySeconds is the minimum number of seconds for which a Machine should be ready before considering the replica available.
+    // Defaults to 0 (machine will be considered available as soon as the Node is ready)
+    // +optional
+    MinReadySeconds int32 `json:"minReadySeconds,omitempty"`
+
+    // If specified, all readiness gates will be evaluated for Machine readiness.
+    // A Machine is ready when `InfrastructureReady`, `NodeHealthy` and `HealthCheckSucceeded` (if present) are "True"; 
+    // if other conditions are defined in this field, those conditions should be "True" as well for the Machine to be ready.
+    // +optional
+    // +listType=map
+    // +listMapKey=conditionType
+    ReadinessGates []MachineReadinessGate `json:"readinessGates,omitempty"`
+
+    // Other fields...
+}
+
+// MachineReadinessGate contains the reference to a Machine condition to be used as readiness gates.
+type MachineReadinessGate struct {
+    // ConditionType refers to a condition in the Machine's condition list with matching type.
+    // Note: Both  Cluster API conditions or conditions added by 3rd party controller can be used as readiness gates.
+    ConditionType string `json:"conditionType"`
+}
+```
+
+| v1beta1 (current)      | v1Beta2 (tentative Q1 2025) | v1beta1 removal (tentative Q1 2026) |
+|------------------------|-----------------------------|-------------------------------------|
+| `ReadinessGates` (new) | `ReadinessGates`            | `ReadinessGates`                    |
+| other fields...        | other fields...             | other fields...                     |
+
+Notes:
+- Both `MinReadySeconds` and `ReadinessGates` should be treated as other in-place propagated fields (changing this should not trigger rollouts).
+- Similarly to Pod's `ReadinessGates`, also Machine's `ReadinessGates` accept only conditions with positive polarity; 
+  The Cluster API project might revisit this in future to stay aligned with Kubernetes or if there are use cases justifying this change.
+
+#### Machine Print columns
+
+| Current           | To be                         |
+|-------------------|-------------------------------|
+| `NAME`            | `NAME`                        |
+| `CLUSTER`         | `CLUSTER`                     |
+| `NODE NAME`       | `PAUSED` (new) (*)            |
+| `PROVIDER ID`     | `NODE NAME`                   |
+| `PHASE` (deleted) | `PROVIDER ID`                 |
+| `AGE`             | `READY` (new)                 |
+| `VERSION`         | `AVAILABLE` (new)             |
+|                   | `UP TO DATE` (new)            |
+|                   | `AGE`                         |
+|                   | `OS-IMAGE` (new) (*)          |
+|                   | `KERNEL-VERSION` (new) (*)    |
+|                   | `CONTAINER-RUNTIME` (new) (*) |
+
```

Comment:
the net addition of 3 extra new columns in the default output could potentially lead to output that frequently spans more that one line in most terminals. See current output below. Now imagine adding 3 more columns:

```
root@421a4016968eb1464e673a6b3ac7bac4 [ ~ ]# kubectl get machine -n tkg-ns tkg-cluster-ujfv-dfcqg-k6wz2
NAME                           CLUSTER            NODENAME                       PROVIDERID                                       PHASE     AGE   VERSION
tkg-cluster-ujfv-dfcqg-k6wz2   tkg-cluster-ujfv   tkg-cluster-ujfv-dfcqg-k6wz2   vsphere://421adb58-19d2-d3c1-9067-ff100888c2ce   Running   28d   v1.29.4+vmware.3-fips.1
```
Do both 'READY' and 'AVAILABLE' need to appear in the default output?

Reply:
In general agree. Our idea was to test for things like this during implementation and then finalize what goes into -o wide.

Do we have a max character number we have in mind as a minimum? Then we can take this, use a "normal" example for names / node names / ... and see that it fits in

Reply:
I have added below all the print columns paragraphs in the proposal the following notes to surface/reinforce what @sbueringer explained above

> Notes:
> - Print columns are not subject to any deprecation rule, so it is possible to iteratively improve print columns without waiting for the next API version.
> - During the implementation we are going to verify the resulting layout and eventually make final adjustments to the column list.

Wrt to `READY` being always visible, the reason why the colums is currently visible is because you can drill down on ready replicas counters from Cluster to Machines (via KCP, MachineDeployments/MachineSet or MachinePool) by using kubectl get; ready replicas counters/ready is consistently visible everywhere.

So, I think that after the final adjustment described above, we either should have `READY`/ready replicas everywhere or drop it everywhere.

Reply:
Agree!

Looking at the example above. Looks like we use a lot of space with information like providerID (and maybe also node name) that could be a lot better used by status information like ready (the ready column takes a lot less space than node name / providerID, in a lot of cases node name is identical to machine name)

Reply:
Resolving because we have agreement on the fact that this requires some fine tuning during/after implementation.
Also tracked in https://github.com/kubernetes-sigs/cluster-api/issues/11105